sw/device/tock/chips/earlgrey/src/chip.rs

```diff
@@ -0,0 +1,269 @@
+//! High-level setup and interrupt mapping for the chip.
+
+use tock::kernel as kernel;
+use tock::rv32i as rv32i;
+
+use core::fmt::Write;
+use core::hint::unreachable_unchecked;
+
+use kernel::common::registers::FieldValue;
+use kernel::debug;
+use rv32i::csr::{mcause, mie::mie, mip::mip, mtvec::mtvec, CSR};
+use rv32i::syscall::SysCall;
+
+use crate::device_config::DEVICE_CONFIG;
+use crate::gpio;
+use crate::interrupts;
+use crate::plic;
+use crate::timer;
+use crate::uart;
+
+pub const CHIP_FREQ: u32 = DEVICE_CONFIG.chip_freq;
+
+pub struct Earlgrey {
+    userspace_kernel_boundary: SysCall,
+    pmp: rv32i::pmp::PMPConfig,
+}
+
+impl Earlgrey {
+    pub unsafe fn new() -> Earlgrey {
+        Earlgrey {
+            userspace_kernel_boundary: SysCall::new(),
+            pmp: rv32i::pmp::PMPConfig::new(4),
+        }
+    }
+
+    pub unsafe fn enable_plic_interrupts(&self) {
+        plic::disable_all();
+        plic::clear_all_pending();
+        plic::enable_all();
+    }
+
+    unsafe fn handle_plic_interrupts() {
+        while let Some(interrupt) = plic::next_pending() {
+            match interrupt {
+                interrupts::UART_TX_WATERMARK..=interrupts::UART_RX_PARITY_ERR => {
+                    uart::UART0.handle_interrupt()
+                }
+                int_pin @ interrupts::GPIO_PIN0..=interrupts::GPIO_PIN31 => {
```

Comment:
If `interrupt` is `Copy`, consider just using that instead of creating a binding here. `var @ pat` is only really useful for non-`Copy` types.